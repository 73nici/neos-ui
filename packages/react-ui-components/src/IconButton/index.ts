import {themr} from '@friendsofreactjs/react-css-themr';
import identifiers from '../identifiers';

import IconButton from './iconButton';

// tslint:disable-next-line:no-var-requires
const style = require('./style.css');

export default themr(identifiers.iconButton, style)(IconButton);
