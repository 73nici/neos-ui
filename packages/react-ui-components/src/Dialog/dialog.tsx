import mergeClassNames from 'classnames';
import {omit} from 'lodash';
import React, {PureComponent, ReactNode} from 'react';
import CloseOnEscape from 'react-close-on-escape';
import Portal from 'react-portal';

import IconButton from '../IconButton';

type DialogStyle = 'wide' | 'narrow';

interface DialogTheme {
    readonly 'dialog': string;
    readonly 'dialog__body': string;
    readonly 'dialog__contentsPosition': string;
    readonly 'dialog__contents': string;
    readonly 'dialog__backDrop': string;
    readonly 'dialog__title': string;
    readonly 'dialog__closeBtn': string;
    readonly 'dialog__actions': string;
    readonly 'dialog--wide': string;
    readonly 'dialog--narrow': string;
}

export interface DialogProps {
    /**
     * This prop controls the rendered state of the Dialog, when falsy, nothing gets rendered into the DOM.
     */
    readonly isOpen: boolean;

    /**
     * The handler which gets called once the user clicks on the close symbol in the top right corner of the Dialog.
     */
    readonly onRequestClose: () => void;

    /**
     * The title to be rendered on top of the Dialogs contents.
     */
    readonly title: ReactNode;

    /**
     * The `style` prop defines the visual style of the `Dialog`.
     */
    readonly style: DialogStyle;

    /**
     * The contents to be rendered within the Dialog.
     */
    readonly children: ReactNode;

    /**
     * An Array of nodes(e.g. Action Buttons) which are placed at the bottom of the Dialog.
     */
    readonly actions: ReadonlyArray<ReactNode>;

    /**
     * An optional `className` to attach to the wrapper.
     */
    readonly className?: string;

    /**
     * An optional `contentsClassName` to attach to the content area of the dialog.
     */
    readonly contentsClassName?: string;

    /**
     * An optional css theme to be injected.
     */
    readonly theme: DialogTheme;
}

export class DialogWithoutEscape extends PureComponent<DialogProps> {
    public render(): JSX.Element {
        const {
            className,
            title,
            style,
            children,
            isOpen,
            onRequestClose,
            actions,
            theme,
            ...restProps
        } = this.props;

        const rest = omit(restProps, ['isOpen']); // TODO: ?

        const finalClassName = mergeClassNames(
            theme.dialog,
            {
                [theme['dialog--wide']]: style === 'wide',
                [theme['dialog--narrow']]: style === 'narrow',
            },
            className,
        );

        const finalClassNameBody = mergeClassNames(
            theme.dialog__body,
            'dialog__body'
        );

        const backDropClassName = mergeClassNames(
            theme.dialog__backDrop,
            'dialog__backDrop'
        );

        return (
            <Portal isOpened={isOpen}>
                <section {...rest} className={finalClassName} role="dialog" tabIndex={0}>
                    <div className={backDropClassName} role="button" onClick={onRequestClose}/>
                    <div className={theme.dialog__contentsPosition}>
                        <div className={theme.dialog__contents}>
                            {onRequestClose && (
                                <IconButton
                                    icon="times"
                                    className={theme.dialog__closeBtn}
                                    onClick={onRequestClose}
                                    size="regular"
                                    style="brand"
                                    hoverStyle="darken"
                                    autoFocus={true}
                                />
                            )}
                            <div className={theme.dialog__title}>
                                {title}
                            </div>

                            <div className={finalClassNameBody}>
                                {children}
                            </div>

                            {actions && actions.length ?
                                <div className={theme.dialog__actions}>
                                    {React.Children.map(actions, (action, index) => <span key={index}>{action}</span>)}
                                </div> : null
                            }
                        </div>
                    </div>
                </section>
            </Portal>
        );
    }
}

// tslint:disable-next-line:max-classes-per-file
class DialogWithEscape extends PureComponent<DialogProps> {
    public render(): JSX.Element {
        return (
            <CloseOnEscape onEscape={this.onEscape}>
                <DialogWithoutEscape {...this.props}/>
            </CloseOnEscape>
        );
    }

    private readonly onEscape = () => {
        this.props.onRequestClose();
    }
}

export default DialogWithEscape;
