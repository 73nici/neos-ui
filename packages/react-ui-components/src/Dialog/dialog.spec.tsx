import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import Portal from 'react-portal';

import IconButton from '../IconButton';
import {DialogProps, DialogWithoutEscape as Dialog} from './dialog';

describe('<Dialog/>', () => {
    const props: DialogProps = {
        actions: ['Foo 1', 'Foo 2'],
        children: 'Foo children',
        isOpen: false,
        onRequestClose: () => null,
        style: 'wide',
        theme: {
            'dialog': 'dialogClassName',
            'dialog--narrow': 'narrowClassName',
            'dialog--wide': 'wideClassName',
            'dialog__actions': 'actionsClassName',
            'dialog__body': 'bodyClassName',
            'dialog__closeBtn': 'closeBtnClassName',
            'dialog__contents': 'contentsClassName',
            'dialog__contentsPosition': 'contentsPositionClassName',
            'dialog__title': 'titleClassName',
        },
        title: 'Foo title',
    };

    it('should render correctly.', () => {
        const wrapper = shallow(<Dialog {...props}/>);

        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should pass a falsy "isOpened" tag to the "Portal" component if the "isOpen" prop is falsy.', () => {
        const wrapper = shallow(<Dialog {...props}/>);
        const portal = wrapper.find(Portal);

        expect(portal.prop('isOpened')).toBe(false);
    });

    it('should pass a truthy "isOpened" tag to the "Portal" component if the "isOpen" prop is truthy.', () => {
        const wrapper = shallow(<Dialog {...props} isOpen={true}/>);
        const portal = wrapper.find(Portal);

        expect(portal.prop('isOpened')).toBe(true);
    });

    it('should render the "dialog--wide" className from the "theme" prop if the style is wide.', () => {
        const wrapper = shallow(<Dialog {...props} style="wide"/>);
        const portal = wrapper.find(Portal);
        const section = portal.find('section');

        expect(section.prop('className')).toContain('wideClassName');
    });

    it('should render the "dialog--narrow" className from the "theme" prop if the style is narrow.', () => {
        const wrapper = shallow(<Dialog {...props} style="narrow"/>);
        const portal = wrapper.find(Portal);
        const section = portal.find('section');

        expect(section.prop('className')).toContain('narrowClassName');
    });

    it('should render the actions if passed.', () => {
        const wrapper = shallow(<Dialog {...props}/>);
        const portal = wrapper.find(Portal);
        const section = portal.find('section');

        expect(section.html().includes('Foo 1')).toBeTruthy();
        expect(section.html().includes('Foo 2')).toBeTruthy();
    });

    it('should call the "onRequestClose" prop when clicking on the "IconButtonComponent" component.', () => {
        const onRequestClose = jest.fn();
        const wrapper = shallow(<Dialog {...props} onRequestClose={onRequestClose}/>);
        const portal = wrapper.find(Portal);
        const btn = portal.find(IconButton);

        btn.simulate('click');

        expect(onRequestClose.mock.calls.length).toBe(1);
    });
});
