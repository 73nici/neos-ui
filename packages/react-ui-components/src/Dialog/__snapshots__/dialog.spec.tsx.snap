// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dialog/> should render correctly. 1`] = `
<Portal
  isOpened={false}
  onClose={[Function]}
  onOpen={[Function]}
  onUpdate={[Function]}
>
  <section
    className="dialogClassName wideClassName"
    role="dialog"
    tabIndex={0}
  >
    <div
      className="contentsPositionClassName"
    >
      <div
        className="contentsClassName"
      >
        <ThemedIconButton
          className="closeBtnClassName"
          composeTheme="deeply"
          hoverStyle="darken"
          icon="times"
          mapThemrProps={[Function]}
          onClick={[Function]}
          size="regular"
          style="brand"
        />
        <div
          className="titleClassName"
        >
          Foo title
        </div>
        <div
          className="bodyClassName dialog__body"
        >
          Foo children
        </div>
        <div
          className="actionsClassName"
        >
          <span
            key="0/.0"
          >
            Foo 1
          </span>
          <span
            key="1/.1"
          >
            Foo 2
          </span>
        </div>
      </div>
    </div>
  </section>
</Portal>
`;
